cmd/kubelet/app/server.go

```diff
@@ -534,7 +534,7 @@ func run(ctx context.Context, s *options.KubeletServer, kubeDeps *kubelet.Depend
 	}
 
 	if kubeDeps.Cloud == nil {
-		if !cloudprovider.IsExternal(s.CloudProvider) {
+		if !cloudprovider.IsExternal(s.CloudProvider) && !utilfeature.DefaultFeatureGate.Enabled(features.DisableCloudProviders) {
```

Comment:
Rather than silently disabling it, I wonder if we should just exit if the gate is enabled and --cloud-provider is set.

Something like:
```go
if utilfeature.DefaultFeatureGate.Enabled(features.DisableCloudProviders) && s.CloudProvider != "" {
   os.Exit(1)
}
```
That logic probably doesn't belong here, but somewhere more generic, maybe in k8s.io/cloud-provider?

Reply:
I'm slightly confused. The k8s.io/cloud-provider is actually meant to be that source binary where things disabled in KAPI, KCM and Kubelet will be running, so there is no point to have this flag there. Or do you suggest it should have an opposite effect, like `EnableCloudProviders`? os.Exit here will kill default kube binaries.

Reply:
Ignore my comment about k8s.io/cloud-provider, I just mean the logic should be in a shared package used by KAPI, KCM and kubelet.

Reply:
`s.CloudProvider` is set to aws, azure, gce, openstack and vsphere still for in-tree providers, and these are the providers the gate should disable. We still need to support "external" going forward